File: farebot-app/src/main/java/com/codebutler/farebot/serialize/gson/CardKeysGsonTypeAdapterFactory.java
package com.codebutler.farebot.serialize.gson;

import android.support.annotation.NonNull;

import com.codebutler.farebot.card.CardType;
import com.codebutler.farebot.card.classic.key.ClassicCardKeys;
import com.codebutler.farebot.key.CardKeys;
import com.google.common.collect.ImmutableMap;
import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.TypeAdapter;
import com.google.gson.TypeAdapterFactory;
import com.google.gson.internal.Streams;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class CardKeysGsonTypeAdapterFactory implements TypeAdapterFactory {

    static final Map<CardType, Class<? extends CardKeys>> CLASSES
            = ImmutableMap.<CardType, Class<? extends CardKeys>>builder()
            .put(CardType.MifareClassic, ClassicCardKeys.class)
            .build();

    @Override
    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
        if (!CardKeys.class.isAssignableFrom(type.getRawType())) {
            return null;
        }
        final Map<CardType, TypeAdapter<CardKeys>> delegates = new HashMap<>();
        for (Map.Entry<CardType, Class<? extends CardKeys>> entry : CLASSES.entrySet()) {
            TypeAdapter<CardKeys> delegateAdapter
                    = (TypeAdapter<CardKeys>) gson.getDelegateAdapter(this, TypeToken.get(entry.getValue()));
            delegates.put(entry.getKey(), delegateAdapter);
        }
        return (TypeAdapter<T>) new CardKeysTypeAdapter(delegates);
    }

    private class CardKeysTypeAdapter extends TypeAdapter<CardKeys> {

        private static final String KEY_CARD_TYPE = "cardType";

        @NonNull private final Map<CardType, TypeAdapter<CardKeys>> mDelegates;

        CardKeysTypeAdapter(Map<CardType, TypeAdapter<CardKeys>> delegates) {
            mDelegates = delegates;
        }

        @Override
        public void write(JsonWriter out, CardKeys value) throws IOException {
            TypeAdapter<CardKeys> delegateAdapter = mDelegates.get(value.cardType());
            JsonObject jsonObject = delegateAdapter.toJsonTree(value).getAsJsonObject();
            Streams.write(jsonObject, out);
        }

        @Override
        public CardKeys read(JsonReader in) throws IOException {
            JsonElement rootElement = Streams.parse(in);
            JsonElement typeElement = rootElement.getAsJsonObject().remove(KEY_CARD_TYPE);
            CardType cardType = CardType.valueOf(typeElement.getAsString());
            TypeAdapter<CardKeys> delegateAdapter = mDelegates.get(cardType);
            return delegateAdapter.fromJsonTree(rootElement);
        }
    }
}

File: farebot-app/src/main/java/com/codebutler/farebot/app/core/serialize/gson/CardKeysGsonTypeAdapterFactory.kt
/*
 * CardKeysGsonTypeAdapterFactory.kt
 *
 * This file is part of FareBot.
 * Learn more at: https://codebutler.github.io/farebot/
 *
 * Copyright (C) 2017 Eric Butler <eric@codebutler.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.codebutler.farebot.app.core.serialize.gson

import com.codebutler.farebot.card.CardType
import com.codebutler.farebot.card.classic.key.ClassicCardKeys
import com.codebutler.farebot.key.CardKeys
import com.google.gson.Gson
import com.google.gson.TypeAdapter
import com.google.gson.TypeAdapterFactory
import com.google.gson.internal.Streams
import com.google.gson.reflect.TypeToken
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import java.util.HashMap

class CardKeysGsonTypeAdapterFactory : TypeAdapterFactory {

    companion object {
        private val KEY_CARD_TYPE = "cardType"

        private val CLASSES = mapOf(
                CardType.MifareClassic to ClassicCardKeys::class.java
        )
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T> create(gson: Gson, type: TypeToken<T>): TypeAdapter<T>? {
        if (!CardKeys::class.java.isAssignableFrom(type.rawType)) {
            return null
        }
        val delegates = HashMap<CardType, TypeAdapter<CardKeys>>()
        for ((key, _) in CLASSES) {
            val cardKeysType = object : TypeToken<CardKeys>() {}
            delegates.put(key, gson.getDelegateAdapter(this, cardKeysType))
        }
        return CardKeysTypeAdapter(delegates) as TypeAdapter<T>
    }

    private inner class CardKeysTypeAdapter internal constructor(
            private val delegates: Map<CardType, TypeAdapter<CardKeys>>) : TypeAdapter<CardKeys>() {

        override fun write(out: JsonWriter, value: CardKeys) {
            val delegateAdapter = delegates[value.cardType()]
                    ?: throw IllegalArgumentException("Unknown type: ${value.cardType()}")
            val jsonObject = delegateAdapter.toJsonTree(value).asJsonObject
            Streams.write(jsonObject, out)
        }

        override fun read(inJsonReader: JsonReader): CardKeys {
            val rootElement = Streams.parse(inJsonReader)
            val typeElement = rootElement.asJsonObject.remove(KEY_CARD_TYPE)
            val cardType = CardType.valueOf(typeElement.asString)
            val delegateAdapter = delegates[cardType]
                    ?: throw IllegalArgumentException("Unknown type: $cardType")
            return delegateAdapter.fromJsonTree(rootElement)
        }
    }
}
